azkaban-common/src/main/java/azkaban/utils/SystemMemoryInfo.java

```diff
@@ -0,0 +1,165 @@
+              freeMemAmount, LOW_MEM_THRESHOLD));
+      return false;
+    }
+
+    if (freeMemAmount - xmx < LOW_MEM_THRESHOLD) {
+      logger.info(String.format("Free memory amount minus xmx (%d - %d kb) is less than low mem threshold (%d kb),  memory request declined.",
+              freeMemAmount, xmx, LOW_MEM_THRESHOLD));
+      return false;
+    }
+
+    if (freeMemDecrAmt > 0) {
+      freeMemAmount -= freeMemDecrAmt;
+    } else {
+      freeMemAmount -= xms;
+    }
+    
+    logger.info("Current free memory amount is " + freeMemAmount);
+    return true;
+  }
+
+  private synchronized static void updateFreeMemAmount(long size) {
+    freeMemAmount = size;
+  }
+
+  private static void readMemoryInfoFile() {
+    BufferedReader br = null;
+    try {
+      br = new BufferedReader(new FileReader(MEMINFO_FILE));
+
+      long sizeMemFree = 0;
+      long sizeBuffers = 0;
+      long sizeCached = 0;
+      long sizeSwapCached = 0;
+      int count = 0;
+      String line = br.readLine();
+      while (line != null) {
+        if (line.startsWith("MemFree:") || line.startsWith("Buffers:")
+            || line.startsWith("Cached") || line.startsWith("SwapCached")) {
+          int idx1 = line.indexOf(":");
+          int idx2 = line.lastIndexOf("kB");
+          String strSize = line.substring(idx1+1, idx2-1).trim();
+          
+          if (line.startsWith("MemFree:")) {
+            sizeMemFree = Long.parseLong(strSize);
+          } else if (line.startsWith("Buffers:")) {
+            sizeBuffers = Long.parseLong(strSize);
+          } else if (line.startsWith("Cached:")) {
+            sizeCached = Long.parseLong(strSize);
+          } else if (line.startsWith("SwapCached:")) {
+            sizeSwapCached = Long.parseLong(strSize);
+          }
+
+          //all lines read
+          if (++count == 4) {
+            break;
+          }
+        }
+        line = br.readLine();
+      }
+
+      if (count < 4) {
+        logger.error("Error: less than 4 rows read from /proc/meminfo for free memory information");
+      }
+
+      long sizeTotal = sizeMemFree + sizeBuffers + sizeCached + sizeSwapCached;
+
+      logger.info(String.format("Current system free memory is %d kb (MemFree %d, Buffers %d, Cached %d, SwapCached %d)",
+              sizeTotal, sizeMemFree, sizeBuffers, sizeCached, sizeSwapCached));
+
+      if (sizeTotal > 0) {
+        updateFreeMemAmount(sizeTotal);
+      }
+    } catch (IOException e) {
+      logger.error("Exception in reading memory info file", e);
+    } finally {
+      try {
+        if (br != null) {
+          br.close();
+        }
+      } catch (IOException e) {
+        logger.error("Exception in closing the buffered reader", e);
+      }
+    }
+  }
+
+  static class MemoryInfoReader implements Runnable {
+    @Override
+    public void run() {
+      readMemoryInfoFile();
```

Comment:
It would be good to add a catch all clause and log out the exception to make it easier to debug.